### ISSUE TITLE
Refactor the models to share building blocks such as projection heads and prediction heads

### ISSUE THREAD

**Issue by IgorSusmelj** [lightly-ai/lightly#232]
We have lots of repetitive code in the various models. Let's take the blocks we can recycle out and put them in a common file such as `building_blocks.py` (feel free to suggest a better name)

Some of the models have slight variations in the heads. E.g. some have batch norm. Some have 3 layers. Others have 2.

Currenlty, for SimCLR we have:
```python
def _get_simclr_projection_head(num_ftrs: int, out_dim: int):
    """Returns a 2-layer projection head.
    Reference (07.12.2020):
    https://github.com/google-research/simclr/blob/master/model_util.py#L141
    """
    modules = [
        nn.Linear(num_ftrs, num_ftrs),
        #nn.BatchNorm1d(num_ftrs),
        nn.ReLU(),
        nn.Linear(num_ftrs, out_dim)
    ]
    return nn.Sequential(*modules)
```

For MoCo we have:
```python
def _get_moco_projection_head(num_ftrs: int, out_dim: int):
    """Returns a 2-layer projection head.
    Reference (07.12.2020):
    https://github.com/facebookresearch/moco/blob/master/moco/builder.py
    """
    modules = [
        nn.Linear(num_ftrs, num_ftrs),
        nn.ReLU(),
        nn.Linear(num_ftrs, out_dim)
    ]
    return nn.Sequential(*modules)
```

**Comment by AdrianArnaiz** [comment]
This issue is very helpful!

* **Projection heads** are very similar between models. It could be moved to another file and code as a parametrizable method to custom the projection head regarding the little differences between models (*Simclr*, *SimSiam* and *Barlow Twins* projection head are slightly different).
*  **Prediction head** is only included in *SimSiam*. However, refactoring this head turns the prediction task into a more flexible feature. It will allow us to combine every combination of `backbone+projection_head+prediction_head`.
* **Forward methods** of the models are also very similar. It seems to me that the "only" differences are regarding the parameters and the usage of the prediction head. 

However, I don't think this last change will actually give more value and functionality to the library. I think this last idea is only related to the code quality, unlike the other 2 ideas that actually provide code quality and functionality.

I'm not currently available to get more issues due to personal workloads. However, I'll definitely take more tickets as soon as I will be able to do it. I'm glad to keep contributing!

**Comment by IgorSusmelj** [comment]
I would also keep the forward methods for now separate. It also makes it easier for others to investigate how the models work. Maybe we don't have to differentiate between projection and prediction head at all for the shared code.

**Comment by philippmwirth** [comment]
See #377, #378, #379

**Comment by MalteEbner** [comment]
closed in favour of #378, because it is a duplicate
